### DISCUSSION TITLE
v1.1.1 versus v1.2.0

### DISCUSSION THREAD

**Discussion by WilliamParsonsV** [CCSDSPy/ccsdspy#101] [Q&A] [answered]
Hello! I've noticed different behavior between versions 1.1.1 (your version history shows only 1.1.0 exists but I have 1.1.1 pip installed) and 1.2.0. 

I have a baseline file created from our existing Perl extract tool and I am comparing it against my Python extract tool that uses ccsdspy. For v1.1.1 I get the assert result that my two dataframes are equivalent for apid=400, yet with **no other changes other than upgrading to ccsdspy==1.2.0** the assertion errors that a few of my columns are now all 0's compared to the baseline which has those columns as all 32 or all 26.

**I've narrowed down the reasoning:**
-  v1.1.1 gives me the error "Big-endian buffer not supported on little-endian compiler" when trying to do a df.drop(columns=) operation for apid=[40, 205]. Yet v1.2.0 gives me this error for apid=[40, 205, and 400] which is the apid I'm using in my tests.

- I get around this error using the below code, but that doesn't seem to actually change the values and therefore doesn't change the assertion result (v1.1.1 = SUCCESS, v1.2.0 = FAIL). But this does raise the question why I get this error for apid=400 only with v1.2.0?
- Worth noting that apid=400 does have polynomial conversions applied to its data

```# Avoid "Big-endian Compiler Error"
                    try:
                        df[apid] = df[apid].drop(CCSDSinfo, axis=1) # Drop CCSDS Info
                    except:
                        for key in df[apid].keys():
                            if df[apid][str(key)].dtype.byteorder == '>':
                                numpy_array = df[apid][str(key)].to_numpy()
                                swapped_array = numpy_array.byteswap().newbyteorder() # Use the byteswap method on the NumPy array
                                df[apid][str(key)] = pd.Series(swapped_array) # Assign the swapped NumPy array back to the DataFrame
    
                        df[apid] = df[apid].drop(CCSDSinfo, axis=1) # Drop CCSDS Info```

**Comment by ddasilva** [comment]
Can you please provide the packet definition for this packet, and some test data? You can email me privately if you don't want to share here. We fixed some issues with signed int handling in 1.2, are the affected fields int or uint?

**Reply by WilliamParsonsV** [reply]
Thanks for getting back to me! I'll confirm with my manager what I'm allowed to send, but for example this is one of the mnemonics that raises the assertion error in apid400:
``` 
"FTS_AOI_STRL_NIR": {
          "description": "AOI Start NIR - Left", 
          "encoding": "U12", 
          "id": 1416, 
          "lengthinbits": 7, 
          "mnemonicid": 1396, 
          "offsetbit": "2", 
          "offsetbyte": "125", 
          "offsetfrom": "STARTOFPACKET", 
          "postpadding": 7, 
          "prepadding": 0, 
          "recordname": "FTS_AOI_STRL_NIR", 
          "recordtype": "PacketMnemonic", 
          "telemetrypacketid": 84} ```

**Reply by WilliamParsonsV** [reply]
It appears they're all unsigned int's

**Reply by ddasilva** [reply]
Great, it sounds like the modification to the uint / int code is the primary suspect. If your manager ends up saying you can't share, please let me know and I'll try some workarounds that don't require you to share data.

**Reply by WilliamParsonsV** [reply]
**Interesting development:** I wrote a supercut version of my extract tool to run with a fakemission.json that just has pkt205 with dummy PARAM names (so that I can share with you). I didn't include conversions because those are tied to the mnemonic names, and I don't get the error 'Big-endian buffer not supported on little-endian compiler'!! Then I run the supercut extract for my_actual_mission.json that has the conversions for pkt205 and I DO get the error for pkt205. So it seems the issue is with the conversion step/converted data

Once I get fakemission.json set-up with conversion info I'll send you apid00205.tlm and fakemission.json, but in the meantime I wanted to update you with my findings

**Reply by WilliamParsonsV** [reply]
Files sent via email to aid in debugging

**Reply by ddasilva** [reply]
Hi @WilliamParsonsV, I created an environment with ccsdspy 1.1.1, and ran the code you gave me. It doesn't look like it's working for me in that version either! Can you check that there's not also different version of pandas, numpy, etc between your functional and non-functional environments?

It seems the line that is a problem is this one:
```python
df[apid] = df[apid].drop(CCSDSinfo, axis=1) # Drop CCSDS Info  #!!! THIS IS WHERE THE ERROR ARISES (ONLY WHEN CONVERSIONS ARE APPLIED)
```

When I rewrite this line as follows, it seems to work. I think this has the same effect as the `drop()` method (remove each column in the `CCSDSinfo` list). Can you confirm? Does this work for you?

```python
for key in CCSDSinfo:
    del df[apid][key]
```

**Reply by WilliamParsonsV** [reply]
Hi @ddasilva! Sorry, I should've been more clear, pkt 205 (the one I gave you) does not work for either 1.1.1 or 1.2.0. Pkt 400 is the apid that works in 1.1.1 and not in 1.2.0, I just couldn't send the info for 400 as its ~10x as many columns that Id have to change the sensitive info for manually and I wanted to get you the data asap. More important than the 1.1.1 vs 1.2.0 I guess is why this error is showing up at all in the first place, regardless of version? It seems the drop function triggers the error but only when conversions are applied, does the code I sent you for applying conversions have any obvious issues? 

Initial testing with the for loop del code you sent above it does work, no errors! Awesome, I'll run my unit level tests with it to confirm the data is also correct with this instead of .drop!

**Reply by ddasilva** [reply]
Great! I'll wait to hear whether this for loop del code works for you. 

I am tempted to say that the error you are getting is more of a pandas issue than a CCSDSPy one (maybe a bug in pandas, or just a 
pandas quirk), and it is surfacing because of something that changed in CCSDSPy that isn't necessarily a CCSDSPy bug. Either way, I'm happy to help out.

**Reply by WilliamParsonsV** [reply]
So it seems the for loop del code works in the sense that it stops the "Big-endian buffer not supported on little-endian compiler" error from occurring, however, running my test_extract.py with v1.1.1 gives No.1 and changing to v1.2.0 and doing nothing else then re-running test_extract.py gives No.2, this is where I was at with my try/except code from the original post above.

1.  Difference b/w BASELINE and ACTUAL: None
2.  Traceback (most recent call last):
  File "c:\Users\wparsons\GMOC\GMOC_SCHM\Testing\test_extract.py", line 55, in <module>
    test_overall_extract(station, orbit, method, tlmFileList=TESTFILE)
  File "c:\Users\wparsons\GMOC\GMOC_SCHM\Testing\test_extract.py", line 53, in test_overall_extract
    print("Difference b/w BASELINE and ACTUAL: " + str(assert_frame_equal(baselineDF, outputDF, check_dtype=False)))
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wparsons\.conda\envs\GMOC_SCHM\Lib\site-packages\pandas\_testing\asserters.py", line 1224, in assert_frame_equal
    assert_series_equal(
  File "C:\Users\wparsons\.conda\envs\GMOC_SCHM\Lib\site-packages\pandas\_testing\asserters.py", line 1014, in assert_series_equal
    _testing.assert_almost_equal(
  File "pandas\_libs\testing.pyx", line 52, in pandas._libs.testing.assert_almost_equal
  File "pandas\_libs\testing.pyx", line 172, in pandas._libs.testing.assert_almost_equal
  File "C:\Users\wparsons\.conda\envs\GMOC_SCHM\Lib\site-packages\pandas\_testing\asserters.py", line 599, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 2] (column name="REMOVING MNEMONIC NAME SORRY") are different

DataFrame.iloc[:, 2] (column name="REMOVING MNEMONIC NAME SORRY") values are different (100.0 %)
[index]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62]
[left]:  [32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32]
[right]: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
At positional index 0, first diff: 32 != 0

**Reply by WilliamParsonsV** [reply]
I agree its probably a Pandas bug but the fact I can get it to specifically occur between 1.1.1 and 1.2.0 means the changes made between those versions might help point to what the Pandas bug actually is

**Reply by ddasilva** [reply]
This is now a different issue, which is that the values changed. This looks like the bugfix we made. I suspect that the new values are correct, and the old ones are not. Can you double check this?

**Reply by WilliamParsonsV** [reply]
Sorry for the slow response, deadlines for other projects, I will double check this today/Monday and get back to you asap

**Comment by Hydravine** [comment]
Okay so after some manual bit-bashing comparing our existing extract tool vs new ccsdspy-based tool I've come back with 2 major findings:

1.  We've been handling 12 bit unsigned, nibble aligned data wrong for 20+ years (only occurred on useless mnemonics) but thank you for bringing this to our attention!
2. For the specific issue I mention above it seems our existing tool is giving the correct result of 32 vs ccsdspy is giving 0. The specific mnemonic spans the byte boundary which may be a case that ccsdpy doesn't handle?
3. This only happens for v1.2.0, not v1.1.1

v1.1.1:
![image](https://github.com/CCSDSPy/ccsdspy/assets/151222967/b5bfc5bc-d1be-4480-92d2-b9af195fa044)

v1.2.0:
![image](https://github.com/CCSDSPy/ccsdspy/assets/151222967/6c1442da-ea36-41b3-8ce6-9e79933110a6)


I've attached here a trimmed definitions .JSON file with the 2 problem parameters as well as the apid400.tlm source pkt file. There are no conversion applied for these 2 mnemonics.
[pkt400trimmed.zip](https://github.com/CCSDSPy/ccsdspy/files/13451675/pkt400trimmed.zip)

**Reply by Hydravine** [reply]
(I recently switched to a Calian GitHub account as my GitHub CoPilot license is through there, but this is still WilliamParsonsV lol)

Let me know if there's anything else I can provide to help debug this!

**Comment by Hydravine** [comment]
Further development! I searched through the diffs of v1.1.1 and 1.2.0 on github (https://github.com/CCSDSPy/ccsdspy/commit/6bc82328d2071fc82b82c38fc7eab98899cdc462) and found this removed code that would appear to solve my issue, so I inserted it back into my v1.2.0 decode.py function and tested, problem solved!!

1.  Why was this code removed between versions?
2.  I'd also suggest slightly altering the arithmetic for the rare scenarios where someone spans more than one byte boundary instead of simply adding =+1 (Ex: could actually span 2 and need =+2)

![decode_removedCode](https://github.com/CCSDSPy/ccsdspy/assets/151222967/190e91a5-afdf-4f79-854b-3d2106866a89)

**Reply by ddasilva** [reply]
Thanks so much! I was able to reproduce your issue, and I pushed a fix in this branch. I had to do something other uncommenting that block (doing so made other tests fail), but you're right in that it was related to that.  I also added some regression tests for this issue. I notice that this bug only affects the `FixedLength` decoder!
https://github.com/CCSDSPy/ccsdspy/tree/parsons-fix

Can you try this branch and confirm it fixes your issue? If so, I will merge it and make a bug fix release.

**Reply by Hydravine** [reply]
Tested and it works! Thank you! Ill stay on this branch until its implemented into the next release.

**Reply by ddasilva** [reply, marked as answer]
Hi @Hydravine @WilliamParsonsV:  the new update (v1.2.1) has been pushed to PyPI. Let me know if there is anything else I can do!
